---
title: "benchmark"
author: "Jason Torres"
date: "11/1/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup 

```{r}

"%&%" <- function(a,b) {paste0(a,b)}
library("dplyr");library("data.table")
library("GenomicRanges")

serv.dir1 <- "/Users/jtorres/FUSE/"
serv.dir2 <- "/Users/jtorres/FUSE5/"

work.dir <- serv.dir2 %&% "projects/mv-compare/"
input.dir <- serv.dir1 %&% "projects/mv-compare/"

mtx.df <- fread(input.dir %&% "input_files/" %&%
          "metaxcan_full_results_filtered_for_model_fit_and_pval_at_0.01.txt")
mtx.df$pairs <- mtx.df$metabolite %&% ":" %&% mtx.df$gene_name

s6.df <- fread(input.dir %&% "input_files/" %&%
                 "shin_supplemetary_tables_tableS6.txt")
s6.df$pairs <- s6.df$metabolite %&% ":" %&% s6.df$gene_name
causal.df <- fread(input.dir %&% "input_files/" %&% "ndunguTabS2.csv")

```

```{r}
causal.sub <- filter(causal.df,!grepl("unknown",predicted_causal))
causal.pairs <- causal.sub$Metaboid%&%":"%&%causal.sub$predicted_causal %>% unique(.) # 61 total but 60 unique 
#                    isobutyrylcarnitine   M33441  SLC22A1          SLC22A1
#                    isobutyrylcarnitine   M33441  SLC22A2          SLC22A1
#
#filter(s6.df,`Predicted causal gene`=="SLC22A2") # Zero 
causal.genes <- causal.sub$predicted_causal %>% unique(.) 

tp.pairs <- causal.pairs[causal.pairs %in% mtx.df$pairs] #%>% length(.)
fn.pairs <- causal.pairs[!(causal.pairs %in% mtx.df$pairs)] #%>% length(.)

```

```{r}

sub.df <- filter(s6.df,`Predicted causal gene`%in% causal.genes) %>% 
  dplyr::select(.,one_of("SNP","Predicted causal gene","Most associated metabolite or ratio"))
write.table(x=sub.df$SNP,work.dir%&%"locus_snps.txt",sep="\t",quote=F,col.names=F,row.names=F)

```

Manually looked up SNP position on Ensembl 37 Bimart (GRCh37.p13) accessed November 1, 2019


```{r}

library("rtracklayer")

mart.df <- fread(work.dir%&%"locus_snps_mart_export.txt") %>% filter(.,!grepl("H",`Chromosome/scaffold name`))
names(mart.df)[1] <- "SNP"

snp.df <- inner_join(sub.df,mart.df,by="SNP")

snp.gr <- GRanges(seqnames = snp.df$`Chromosome/scaffold name`,
                  IRanges(snp.df$`Chromosome/scaffold position start (bp)`,
                          snp.df$`Chromosome/scaffold position end (bp)`))

pth <- serv.dir2 %&% "datasets/gencode.v30lift37.annotation.gtf"
genes.df <- rtracklayer::import(pth)



```



# functions 

```{r}

# Estimate True Negatives 
est.tn <- function(total.genes,tp,fp,fn){
  total.genes - (tp + fp) - fn 
}

# Estimate Specificity 
est.spec <- function(tn,fp){ 
  tn/(tn+fp)
}

# Estimate Accuracy 
est.acc <- function(tn,tp,fn,fp){
  (tn+tp)/(tn+tp+fn+fp)
}

```

```{r}


tp <- 41 # true positives
fn <- 19 # false negatives; not 20 it seems! Must correct in manuscript 
fp <- 473 # false positives 

tn.est1 <- est.tn(1000,tp,fp,fn)
spec.est1 <- est.spec(tn.est1,fp)
acc.est1 <- est.acc(tn.est1,tp,fn,fp)

```

